### PR TITLE
🐛  Destination BigQuery: Introduced chunk size config

### LINKED ISSUE #3549: Destination Bigquery returns 404 when uploading data + resumable
## Expected Behavior
Sync between Shopify and BQ works

## Current Behavior
It's not totally clear why this happens... 
BigQuery returns 404 when uploading data after running a few hours.

From Google Cloud API Page [here](https://cloud.google.com/bigquery/docs/reference/api-uploads#best-practices)
> Handle 404 Not Found and 410 Gone errors when doing resumable uploads by starting the entire upload over from the beginning.

## Logs
```
2021-05-21 03:10:39 ERROR (/tmp/workspace/6/0) LineGobbler(voidCall):69 - Exception in thread "main" com.google.cloud.bigquery.BigQueryException: 404 Not Found
2021-05-21 03:10:39 ERROR (/tmp/workspace/6/0) LineGobbler(voidCall):69 - PUT https://www.googleapis.com/upload/bigquery/v2/projects/earthbones-shopify/jobs?uploadType=resumable&upload_id=ABg5-UyhG8f29jrwEfJJ2mu4JTAIjUF_B3F5JvymoctXdtBQjQ29LsavMFtadgBiXThGjQcTtZAffIZJTf0940HdexnlUOtGew
2021-05-21 03:10:39 ERROR (/tmp/workspace/6/0) LineGobbler(voidCall):69 - 	at com.google.cloud.bigquery.spi.v2.HttpBigQueryRpc.translate(HttpBigQueryRpc.java:115)
2021-05-21 03:10:39 ERROR (/tmp/workspace/6/0) LineGobbler(voidCall):69 - 	at com.google.cloud.bigquery.spi.v2.HttpBigQueryRpc.write(HttpBigQueryRpc.java:705)
2021-05-21 03:10:39 ERROR (/tmp/workspace/6/0) LineGobbler(voidCall):69 - 	at com.google.cloud.bigquery.TableDataWriteChannel$1.call(TableDataWriteChannel.java:59)
2021-05-21 03:10:39 ERROR (/tmp/workspace/6/0) LineGobbler(voidCall):69 - 	at com.google.cloud.bigquery.TableDataWriteChannel$1.call(TableDataWriteChannel.java:54)
2021-05-21 03:10:39 ERROR (/tmp/workspace/6/0) LineGobbler(voidCall):69 - 	at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:105)
2021-05-21 03:10:39 ERROR (/tmp/workspace/6/0) LineGobbler(voidCall):69 - 	at com.google.cloud.RetryHelper.run(RetryHelper.java:76)
2021-05-21 03:10:39 ERROR (/tmp/workspace/6/0) LineGobbler(voidCall):69 - 	at com.google.cloud.RetryHelper.runWithRetries(RetryHelper.java:50)
2021-05-21 03:10:39 ERROR (/tmp/workspace/6/0) LineGobbler(voidCall):69 - 	at com.google.cloud.bigquery.TableDataWriteChannel.flushBuffer(TableDataWriteChannel.java:53)
2021-05-21 03:10:39 ERROR (/tmp/workspace/6/0) LineGobbler(voidCall):69 - 	at com.google.cloud.BaseWriteChannel.flush(BaseWriteChannel.java:112)
2021-05-21 03:10:39 ERROR (/tmp/workspace/6/0) LineGobbler(voidCall):69 - 	at com.google.cloud.BaseWriteChannel.write(BaseWriteChannel.java:139)
2021-05-21 03:10:39 ERROR (/tmp/workspace/6/0) LineGobbler(voidCall):69 - 	at io.airbyte.integrations.destination.bigquery.BigQueryDestination$RecordConsumer.acceptTracked(BigQueryDestination.java:358)
2021-05-21 03:10:39 ERROR (/tmp/workspace/6/0) LineGobbler(voidCall):69 - 	at io.airbyte.integrations.base.FailureTrackingAirbyteMessageConsumer.accept(FailureTrackingAirbyteMessageConsumer.java:66)
2021-05-21 03:10:39 ERROR (/tmp/workspace/6/0) LineGobbler(voidCall):69 - 	at io.airbyte.integrations.base.IntegrationRunner.consumeWriteStream(IntegrationRunner.java:132)
2021-05-21 03:10:39 ERROR (/tmp/workspace/6/0) LineGobbler(voidCall):69 - 	at io.airbyte.integrations.base.IntegrationRunner.run(IntegrationRunner.java:113)
2021-05-21 03:10:39 ERROR (/tmp/workspace/6/0) LineGobbler(voidCall):69 - 	at io.airbyte.integrations.destination.bigquery.BigQueryDestination.main(BigQueryDestination.java:421)
```

## Steps to Reproduce
It's not clear how to reproduce. Maybe this happens when trying to sync large Shopify datasets to BQ.

## Severity of the bug for you
Very low / Low / **Medium** / High / Critical

## Airbyte Version
0.23.0-alpha

## Connector Version (if applicable)
0.2.2 Shopify

## Additional context
Slack [convo](https://airbytehq.slack.com/archives/C01MFR03D5W/p1621510734082900)

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1200368060087111/1200368172141330) by [Unito](https://www.unito.io)

### PR DESCRIPTION
**Pull request by etsybaev**
## What
As part of the ticket https://github.com/airbytehq/airbyte/issues/3549 I've found that for big datasets we may fail with "out of memory" error. This may happen when someone wants to migrate a big data sets especially with lots of tables\streams.  That is caused due to the fact that used Google SDK BigQuery client that is created for each table by default uses 15MiB for buffer in HEAP memory. Usually, JVM allocates about 25% of all available memory for HEAP.  So big datasets destination container crashes with out of memory error and doesn't properly notify the web app as non of "autoclosable" methods called.

## How
Introduced a new config param. So now user may decrease this value. That would cause more requests to google big query server but may save lots of memory and prevent crashes and freezes. 
This is how new arg looks like ou UI:
Selection_274

## Pre-merge Checklist
Expand the checklist which is relevant for this PR. 

<details><summary> <strong> Connector checklist </strong> </summary>
<p>

- [ ] Issue acceptance criteria met
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Secrets are annotated with `airbyte_secret` in the connector's spec
- [ ] Credentials added to Github CI if needed and not already present. [instructions for injecting secrets into CI](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#using-credentials-in-ci). 
- [x] Unit & integration tests added as appropriate (and are passing)
    * Community members: please provide proof of this succeeding locally e.g: screenshot or copy-paste acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [x] `/test connector=connectors/<name>` command as documented [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-an-existing-connector) is passing. 
    * Community members can skip this, Airbyters will run this for you. 
- [x] Code reviews completed
- [ ] Documentation updated 
    - [ ] `README.md`
    - [ ] `docs/SUMMARY.md` if it's a new connector
    - [x] Created or updated reference docs in `docs/integrations/<source or destination>/<name>`.
    - [x] Changelog in the appropriate page in `docs/integrations/...`. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md` contains a reference to the new connector
    - [ ] Build status added to [build page](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/builds.md)
- [x] Build is successful
- [x] Connector version bumped like described [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-a-connector)
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-a-connector)
- [x] No major blockers
- [ ] PR merged into master branch
- [ ] Follow up tickets have been created
- [ ] Associated tickets have been closed & stakeholders notified
</p>
</details>

<details><summary> <strong> Connector Generator checklist </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

**Review by yaroslav-hrytsaienko — APPROVED** [review]

**Comment by etsybaev** [comment]
Hi @sherifnada @subodh1810 @tuliren.
This PR introduces a chunk\buffer size config for big query SDK client. I would propose to add it at least for a while and ask those who have an issue with migration to decrease the buffer size and increase the allocated RAM memory. Just to see it helps. If yes - then we would definitely know the reason. If now - we may think about rolling back this PR

**Review by subodh1810 — CHANGES_REQUESTED** [review]
This will break existing connectors cause none of them would have chunk size selected. Also what about handling 404 and adding more logs and specifying the job location and project like we discussed in the meeting.

**Comment by etsybaev** [comment]
> This will break existing connectors cause none of them would have chunk size selected. Also what about handling 404 and adding more logs and specifying the job location and project like we discussed in the meeting.

Hi @subodh1810.

I've added more logs as part of this PR and here described the point about 404 https://github.com/airbytehq/airbyte/pull/5049#discussion_r678585179


As an option to keep the backward compatibility I may update this PR making this new arg as an optional. If not set - then will use the default BigQuery SDK client's value. What do you think?

**Comment by subodh1810** [comment]
Ahan! alright! I think its important to keep this as backward compatible and make a default choice if the value is not present

**Comment by etsybaev** [comment]
Hi @subodh1810. I've updated the PR -> removed new arg from the mandatory list. Will use default SDK BigQuery client's values if other is not set explicitly in config

File: airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDestination.java [hunk 1]
```diff
@@ -77,6 +77,8 @@
   static final String CONFIG_PROJECT_ID = "project_id";
   static final String CONFIG_DATASET_LOCATION = "dataset_location";
   static final String CONFIG_CREDS = "credentials_json";
+  static final String BIG_QUERY_CLIENT_CHUNK_SIZE = "big_query_client_buffer_size_mb";
+  final int MiB = 1024 * 1024;
```
Comment (subodh1810) [inline airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDestination.java:81]: `private static final int MIB = 1024 * 1024;`

**Review by subodh1810 — APPROVED** [review]
Just 1 comment, rest looks fine

File: airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDestination.java (hunk 1 above)
Comment (etsybaev) [reply airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDestination.java]: Updated, thanks

**Pull request merged**